core/controllers/suggestion.py

```diff
@@ -0,0 +1,148 @@
+
+"""Controllers for suggestions."""
+
+from constants import constants
+from core.controllers import base
+from core.domain import acl_decorators
+from core.domain import suggestion_services
+from core.platform import models
+
+(suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
+
+
+class SuggestionHandler(base.BaseHandler):
+    """"Handles operations relating to learner suggestions."""
+
+    @acl_decorators.can_suggest_changes
+    def post(self):
+        if not constants.USE_NEW_SUGGESTION_FRAMEWORK:
+            raise self.PageNotFoundException
+
+        suggestion_services.create_suggestion(
+            self.payload.get('suggestion_type'),
+            self.payload.get('target_type'), self.payload.get('target_id'),
+            self.payload.get('target_version_at_submission'),
+            self.user_id, self.payload.get('change_cmd'),
+            self.payload.get('description'),
+            self.payload.get('assigned_reviewer_id'),
+            self.payload.get('final_reviewer_id'))
+        self.render_json(self.values)
+
+
+class SuggestionToExplorationActionHandler(base.BaseHandler):
+    """Handles actions performed on suggestions to explorations."""
+
+    ACTION_TYPE_ACCEPT = 'accept'
+    ACTION_TYPE_REJECT = 'reject'
+
+
+    # TODO (nithesh): Add permissions for users with enough scores to review
+    # Will be added as part of milestone 2 of the generalized review system
+    # project.
+    @acl_decorators.can_edit_exploration
+    def put(self, exploration_id, suggestion_id):
+        if not constants.USE_NEW_SUGGESTION_FRAMEWORK:
+            raise self.PageNotFoundException
+
+        if len(suggestion_id.split('.')) != 3:
+            raise self.InvalidInputException('Invalid format for suggestion_id.'
+                                             ' It must contain 3 parts'
+                                             ' separated by \'.\'')
+
+        if suggestion_id.split('.')[0] != 'exploration':
+            raise self.InvalidInputException('This handler allows actions only'
+                                             ' on suggestions to explorations.')
+
+        if suggestion_id.split('.')[1] != exploration_id:
+            raise self.InvalidInputException('The exploration id provided does '
+                                             'not match the exploration id '
+                                             'present as part of the '
+                                             'suggestion_id')
+
+        action = self.payload.get('action')
+        if action == self.ACTION_TYPE_ACCEPT:
+            suggestion = suggestion_services.get_suggestion_by_id(suggestion_id)
+            suggestion_services.accept_suggestion(
+                suggestion, self.user_id, self.payload.get('commit_message'),
+                self.payload.get('review_message'))
+        elif action == self.ACTION_TYPE_REJECT:
+            suggestion = suggestion_services.get_suggestion_by_id(suggestion_id)
```

Comment:
We should move line 84 and 79 out of the `if..elif` block.

Reply:
I thought so. Then I thought if the action is invalid, we could save a db query so didn't move it outside. I don't think there will be too much difference though, I'll move it out.

Reply:
Done